src/Orleans/Core/Grain.cs

```diff
@@ -332,5 +324,51 @@ protected virtual Task ReadStateAsync()
         {
             return storage.ReadStateAsync();
         }
+
+        void ILifecycleParticipant<GrainLifecycleStage>.Participate(ILifecycleObservable<GrainLifecycleStage> lifecycle)
+        {
+            lifecycle.Subscribe(GrainLifecycleStage.SetupState, this);
+        }
+
+        async Task ILifecycleObserver.OnStart(CancellationToken ct)
```

Comment:
maybe the `ILifecycleObserver` itself can be instantiated on the `Participate` phase, so that the grain itself doesn't have to implement `ILifecycleObserver` that could conflict with how the users implement their grains?
For example the `StateStorageBridge` itself could implement that interface, or an inner class of it, or even have it implicitly implemented with a lambda that can use closures.